src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +558,109 @@ ApplicationImpl::joinAllThreads()
     LOG(DEBUG) << "Joined all " << mWorkerThreads.size() << " threads";
 }
 
-bool
-ApplicationImpl::manualClose()
+std::string
+ApplicationImpl::manualClose(optional<uint32_t> const& manualLedgerSeq,
+                             optional<TimePoint> const& manualCloseTime)
 {
     // Manual close only makes sense for validating nodes
     if (!mConfig.NODE_IS_VALIDATOR)
     {
-        return false;
+        throw std::invalid_argument(
+            "Issuing a manual ledger close requires NODE_IS_VALIDATOR to be "
+            "set to true.");
     }
 
     if (mConfig.MANUAL_CLOSE)
     {
-        mHerder->triggerNextLedger(mLedgerManager->getLastClosedLedgerNum() + 1,
-                                   true);
-        return true;
+        auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
+        if (startLedgerSeq == std::numeric_limits<uint32_t>::max())
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Manually closed ledger "
+                                       "sequence number already at max ({})"),
+                            startLedgerSeq));
+        }
+        auto minLedgerSeq = startLedgerSeq + 1;
+        auto const targetLedgerSeq =
+            manualLedgerSeq ? *manualLedgerSeq : minLedgerSeq;
+        if (targetLedgerSeq < minLedgerSeq)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close ledger "
+                                       "sequence number {} (minimum {})"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        if (!mHerder->isTransactionQueueEmpty() &&
+            targetLedgerSeq != minLedgerSeq)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("May not skip ledgers (requested "
+                                       "sequence number {}; next {}) while "
+                                       "transaction queue is non-empty"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        auto nextLedgerSeqToTrigger = startLedgerSeq;
+        bool lastClose;
+        do
```

Comment:
This loop seems overly complicated, why not just closing one ledger like before?

I think all you need to make this work is ensure that we meet the conditions so that "herder" can trigger the ledger properly (what I mentioned twice already in different comments).

Glancing at the code, I think that all you need is to call `restoreSCPState` with the right `index`/`StellarValue` when you modify lcl (NB: from `StellarValue`, we only use the close time).

Reply:
>This loop seems overly complicated, why not just closing one ledger like before?

> I think all you need to make this work is ensure that we meet the conditions so that "herder" can trigger the ledger properly (what I mentioned twice already in different comments).

It was because I hadn't figured out how to meet the conditions to skip ledger numbers yet.  I wanted to see if I could get _something_ working first.

> Glancing at the code, I think that all you need is to call restoreSCPState with the right index/StellarValue when you modify lcl (NB: from StellarValue, we only use the close time).

Currently I'm taking Jon's suggestion of using (a slight variation on) `LedgerManager::advanceLedgerPointers()`, and thus far it appears to be working, but I can try the `restoreSCPState()` approach if that's better (maybe with just `advanceLedgerPointers()` there's something that's supposed to be happening that isn't?).

Reply:
> Glancing at the code, I think that all you need is to call restoreSCPState with the right index/StellarValue when you modify lcl (NB: from StellarValue, we only use the close time).

Although some tests were working before using just Jon's `advanceLedgerPointers()` suggestion, there is at least this interface-violating way of calling `restoreSCPState()` as well, in case that is also necessary:  c23a2d05d370d2a67ff4e43cf3c33df9b9b6eb2c

For now, I've left that in a separate commit for independent review, since I'm not sure whether it's necessary or correct, and even if it's both, it should probably be exposed via some new Herder-interface abstract method, though I'm not sure with exactly what parameters or what to call it.